File: src/main/java/examplejava/OptionalExample.java
package examplejava;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;
import java.util.stream.Stream;

public class OptionalExample {
    public static void main(String[] args) {
        //Creating Optionals
        Optional<String> name = Optional.of("Hello");

        //Transforming
        name.map(String::toUpperCase);

        //Filter non empty from list of Optionals
        Stream.of(name, Optional.empty())
                .filter(Optional::isPresent)
                .map(Optional::get)
                .forEach(System.out::println); //Prints Hello

        //Do something only if not null
        name.ifPresent(System.out::println);

        //Default values in case of null
        int length = name.map(String::length).orElse(0);

        //Throw exception in case of null
        length = name.map(String::length).orElseThrow(() -> new IllegalArgumentException("Can't be null"));
    }

    public static String javaMethodWithNullable(String nullableInput,
                                                @NotNull  String nonNullableAnnotatedInput ,
                                                @Nullable String nullableAnnotatedInput){
        return nullableInput.toUpperCase() + nonNullableAnnotatedInput.toUpperCase() + nullableAnnotatedInput.toUpperCase();
    }


}

File: src/main/kotlin/OptionalExample.kt
import examplejava.OptionalExample
import examplejava.OptionalExample.javaMethodWithNullable
import java.util.*

fun main(args: Array<String>) {

    //Creating Optionals
    val name : String? = "Hello"

    //Transforming
    name?.toUpperCase();

    //Filter non empty from list of Optionals
    println(listOf(name, null).filterNotNull()) //prints [Hello]

    //Do something only if not null
    name.let { println(it) }

    //Default values in case of null
    var length = name?.length ?: 0

    //Throw exception in case of null
    length = name?.length  ?: throw IllegalArgumentException("Can't be null")

    callJavaNullableFunction();
}

fun callJavaNullableFunction(){

    //Will throw NPE since we are passing null value to Java parameter that is not annotated.
    var result = javaMethodWithNullable(null, "hello", "world")

    //Compilation error Null can't be value for non-null type String since we are passing null to Java parameter annotated as Not Null
    //result = javaMethodWithNullable("Hello", null, "world")

    //Compiles fine as we are passing null to Java parameter annotated as Nullable. But will result in NPE at runtime.
    result = javaMethodWithNullable("Hello", "World", null)

    println(result)
}